### ISSUE TITLE
[BlenderBIM] View is placed outside of border on sheet

### ISSUE THREAD

**Issue by globalcitizen** [IfcOpenShell/IfcOpenShell#882]
As per the following screenshot from Inkscape.

<img width="681" alt="Screen Shot 2020-06-05 at 21 38 03" src="https://user-images.githubusercontent.com/505366/83872354-2c08ae00-a775-11ea-934a-d645469b8b1c.png">

**Comment by Moult** [comment]
At the moment, there is no way of knowing the margin of the titleblock template, so it just places all new views on the top left corner. The usual workflow, for any CAD/BIM app is for the user to organise the views on a sheet how they see fit. I have left this exercise for the user, in an app such as Inkscape.

Perhaps a better solution to aid this process is to:

 1. Establish a convention of margin in the titleblock SVG, which can be parsed
 2. Use a packing algorithm to place views logically.

Thoughts? Would you like to try and build it?

**Comment by globalcitizen** [comment]
Thinking about it further I guess there are really a few issues here.

- Currently there is no way to tell what the border is.
   - This could be fixed by establishing an explicit ID that can be assigned to an object within the SVG like "drawing_contents". The object itself could be a blank/invisible box or some sort of group element with an offset.
- However, there is both a margin from the border and a border itself (rarely do you want the drawing to touch the border)
   - This could be adaptive (eg. based on the number and size of the views placed on the sheet) or use an explicit minimum.

Before diving in and adding features I would like to get more of a sense of what is present and have some more time to test and comprehend it vs. my urgent requirements, also it would be good to see what ideas others have in the meeting. I will be happy to contribute new code once I have a better idea what's going on!

**Comment by Moult** [comment]
No worries. Another easy fix would be simply to insert it into the center of the sheet. The user then moves it in Inkscape to however they want to organise it.

Unless there is a better packing algorithm you can suggest. The problem with a packing algorithm is that it may not work too well once the user has already placed a series of views, and a new view is added.

**Comment by globalcitizen** [comment]
When I learned technical drawing there was a preference for placing various orthographic projections in relative space around one another, ie.

```
                           [top plan view]
[left elevation]   [front elevation]   [right elevation]
```

... such that the features visible on each individual rendering aligned for visual reference and clarity. That would be one feature to consider longer term.

**Comment by Moult** [comment]
That is correct, but the add-on currently has no way of knowing a plan view from a front elevation.

**Comment by globalcitizen** [comment]
The 'Lighting: Sun position' add-on seems to have no problem assuming Z is up, therefore this seems a sane assumption.

**Comment by Moult** [comment]
Oh yes, you're absolutely right. That's a smart thing to do.

From experience, however, there are many "standard" layouts. What you've described is common if the building is small and fits on one page, but is rarely the case on larger projects. Also, I'd say 99% of drawings are not an overall plan/elevations, and so that pattern doesn't apply.

I do know, however, that when people have gone the route of automating view placements on sheets, often they offer a set of preset layout patterns, and let people define the mapping of pattern to view types themselves. Layouts could be things like top half bottom half, or left half, and top right, bottom right, etc.

Because all of this is a bit further down the road, I'd settle for a simple arbitrary packing first, and focus on other higher priorities. I'm not closing this bug, though, as it is very valid.

**Comment by globalcitizen** [comment]
> What you've described is common if the building is small and fits on one page, but is rarely the case on larger projects [...] 99% of drawings are not an overall plan/elevations.

I agree and of course defer to your experience regarding large projects. However, for 'normal' projects - and I have actually read through in full about 30x development applications in Sydney over the last week which probably represent a fair sampling of small scale commercial and  residential development - my conclusion is that they mostly provide precisely a subset (rarely all) of these drawings, then a few extra drawings such as special purpose plan views (site study, drainage plans, shadow plans, water retention feature construction diagrams, etc.).

The goal would be only a reasonable default - a fair set of assumptions to help most people enhance usability in the normal cases.

Basically I propose assuming:
- If there is a configured (site is geopositioned + oriented) 'Lighting: Sun Position' add-on then we should attempt to use the cardinal points.
- Otherwise, there is a 'front' and we can use the terms front, rear, left, right.

The user should be able to manually specify drawing names anyway.

The above should ideally be internationalized.

I have moved these concerns to dedicated bugs - feature 'auto-name views' (#890) and bug 'strings should be i18n' (#889).
